intel_adsp: ace: clock gating in idle

tmleman opened the pull request:
This patch enables DSP clock gating for ACE platforms. By default, clock gating is blocked by the firmware in the hardware configuration. If CONFIG_ADSP_IDLE_CLOCK_GATING is enabled, this prevent is not active and clock can be gated when core is in idle state. WIth this option disabled clock gating will only be enabled in hardware during power gating.

ceolin commented:
Why don't you have shallow sleep states, then in `pm_state_set` you do what you are doing in `pm_idle_entry` and calls `k_cpu_idle`. Then in the post ops you do what you are doing in `pm_idle_exit` ? The question is if that was considered

ceolin commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -36,7 +36,21 @@
+				substate-id = <2>;
+				min-residency-us = <0>;
+				exit-latency-us = <0>;
+			};
```
how these two states with the same residency time are supposed to be used ? The way it is now, only one will be selected ... unless it has a constraint

tmleman replied:
I updated the description, please take a look

ceolin replied:
this is still something that is not making sense for me. This is looking to me that you are creating this states just to set a property that you want to be used next time the cpu goes idle. These states are never used by the power management subsystem. The only usage is that you set a lock in one of them and check it when the system is goes idle (some state `!PM_STATE_ACTIVE`). It is much cleaner (and less costly) if you add and soc specific API since that is particular to this target (not even part of the SOF fw).

tmleman commented:
> Why don't you have shallow sleep states, then in `pm_state_set` you do what you are doing in `pm_idle_entry` and calls `k_cpu_idle`. Then in the post ops you do what you are doing in `pm_idle_exit` ? The question is if that was considered

If I added this as a sub-state for PM_STATE_RUNTIME_IDLE for example, PM_STATE_ACTIVE would become a state that we would probably end up in less than 5% of the time.

I also think it wouldn't be very efficient. We gain the most when we switch the clock and wait for the next interrupt as long as possible.

tmleman commented:
Update ww45: I will rewrite it using the changes made in https://github.com/zephyrproject-rtos/zephyr/pull/64760. After that ACE will have its own idle implementation, so I dont need to add anything to the kernel idle.

marc-hb commented on dts/xtensa/intel/intel_adsp_ace20_lnl.dtsi:
While you're working on this file, could you please help with the 120 lines of already existing warnings caused by it? They drown anything that could be caused by this PR.

(In a separate PR of course)

```
build-lnl/zephyr/zephyr.dts:61.20-65.5: Warning (unique_unit_address_if_enabled): /soc/alh0@24400: duplicate unit-address (also used in node /soc/alh7@24400)
build-lnl/zephyr/zephyr.dts:66.20-70.5: Warning (unique_unit_address_if_enabled): /soc/alh1@24400: duplicate unit-address (also used in node /soc/alh7@24400)
build-lnl/zephyr/zephyr.dts:71.20-75.5: Warning (unique_unit_address_if_enabled): /soc/alh2@24400: duplicate unit-address (also used in node /soc/alh7@24400)
build-lnl/zephyr/zephyr.dts:76.20-80.5: Warning (unique_unit_address_if_enabled): /soc/alh3@24400: duplicate unit-address (also used in node /soc/alh7@24400)
build-lnl/zephyr/zephyr.dts:81.20-85.5: Warning (unique_unit_address_if_enabled): /soc/alh4@24400: duplicate unit-address (also used in node /soc/alh7@24400)
build-lnl/zephyr/zephyr.dts:86.20-90.5: Warning (unique_unit_address_if_enabled): /soc/alh5@24400: duplicate unit-address (also used in node /soc/alh7@24400)
build-lnl/zephyr/zephyr.dts:91.20-95.5: Warning (unique_unit_address_if_enabled): /soc/alh6@24400: duplicate unit-address (also used in node /soc/alh7@24400)
build-lnl/zephyr/zephyr.dts:61.20-65.5: Warning (unique_unit_address_if_enabled): /soc/alh0@24400: duplicate unit-address (also used in node /soc/alh8@24400)
build-lnl/zephyr/zephyr.dts:66.20-70.5: Warning (unique_unit_address_if_enabled): /soc/alh1@24400: duplicate unit-address (also used in node /soc/alh8@24400)
build-lnl/zephyr/zephyr.dts:71.20-75.5: Warning (unique_unit_address_if_enabled): /soc/alh2@24400: duplicate unit-address (also used in node /soc/alh8@24400)
build-lnl/zephyr/zephyr.dts:76.20-80.5: Warning (unique_unit_address_if_enabled): /soc/alh3@24400: duplicate unit-address (also used in node /soc/alh8@24400)
build-lnl/zephyr/zephyr.dts:81.20-85.5: Warning (unique_unit_address_if_enabled): /soc/alh4@24400: duplicate unit-address (also used in node /soc/alh8@24400)
build-lnl/zephyr/zephyr.dts:86.20-90.5: Warning (unique_unit_address_if_enabled): /soc/alh5@24400: duplicate unit-address (also used in node /soc/alh8@24400)
build-lnl/zephyr/zephyr.dts:91.20-95.5: Warning (unique_unit_address_if_enabled): /soc/alh6@24400: duplicate unit-address (also used in node /soc/alh8@24400)
build-lnl/zephyr/zephyr.dts:96.20-100.5: Warning (unique_unit_address_if_enabled): /soc/alh7@24400: duplicate unit-address (also used in node /soc/alh8@24400)
build-lnl/zephyr/zephyr.dts:61.20-65.5: Warning (unique_unit_address_if_enabled): /soc/alh0@24400: duplicate unit-address (also used in node /soc/alh9@24400)
build-lnl/zephyr/zephyr.dts:66.20-70.5: Warning (unique_unit_address_if_enabled): /soc/alh1@24400: duplicate unit-address (also used in node /soc/alh9@24400)
build-lnl/zephyr/zephyr.dts:71.20-75.5: Warning (unique_unit_address_if_enabled): /soc/alh2@24400: duplicate unit-address (also used in node /soc/alh9@24400)
build-lnl/zephyr/zephyr.dts:76.20-80.5: Warning (unique_unit_address_if_enabled): /soc/alh3@24400: duplicate unit-address (also used in node /soc/alh9@24400)
build-lnl/zephyr/zephyr.dts:81.20-85.5: Warning (unique_unit_address_if_enabled): /soc/alh4@24400: duplicate unit-address (also used in node /soc/alh9@24400)
build-lnl/zephyr/zephyr.dts:86.20-90.5: Warning (unique_unit_address_if_enabled): /soc/alh5@24400: duplicate unit-address (also used in node /soc/alh9@24400)

```

marc-hb replied:
- https://github.com/thesofproject/sof/issues/8356

tmleman commented:
Update ww49 (1/2): dropping dynamic clock switching implementation (it will be added in separate PR)

tmleman commented:
Update ww49 (2/2): rebase.

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -165,7 +171,7 @@ void power_gate_entry(uint32_t core_id)
 	soc_cpus_active[core_id] = false;
 	sys_cache_data_flush_range(soc_cpus_active, sizeof(soc_cpus_active));
-	k_cpu_idle();
+	wait_for_interrupt(0);
```
This is not needed if you always do `waiti 0`. Just call `k_cpu_idle()` that it will end up calling your `arch_cpu_idle()` function. All state conditions could be checked there then ...

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -373,4 +379,22 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+		DSPCS.bootctl[arch_proc_id()].bctl |= DSPBR_BCTL_WAITIPCG;
+	} else {
+		DSPCS.bootctl[arch_proc_id()].bctl &= ~DSPBR_BCTL_WAITIPCG;
+	}
```
Another thing, you are just checking substate 1 here, and for all other states / substates you do the same thing, so why have you declared two substates ?

kv2019i replied:
@ceolin AFAIU, the other substate is part of https://github.com/zephyrproject-rtos/zephyr/pull/66092 which has additional patches on top of this series but DNM for now as it requires more work. We'd like to move this dynamic clock gating part (that is in this PR) faster to upstream and it has proven to work reliably and brings immediate power savings.

ceolin replied:
that ok, I just don't think using active power state + locking this state is the right way to do this. In the end you just want some variable to tell you what to do when you get idle. I mean, the way it is implemented now works, but it is more complicated than it should be and it is using the API / subsystem in a way that was not intended to be used.

Where this lock is being set btw ?

tmleman replied:
> Where this lock is being set btw ?

Here you can find lock usage on SOF side https://github.com/thesofproject/sof/pull/8423

ceolin reviewed (changes requested):
Please take a look in individual comments.

tmleman commented:
@ceolin is the use for substates for PM_STATE_ACTIVE a blocker for you?

ceolin commented:
> @ceolin is the use for substates for PM_STATE_ACTIVE a blocker for you?

I think that is the wrong way to do this, that is not the PM_STATE_ACTIVE semantic. In the end what we need is a way to set a CPU property to when it gets idle decides how to set its clock. Right ?

tmleman committed:
intel_adsp: power: clock gating in idle

This patch enables DSP clock gating for ACE platforms. By default, clock
gating is blocked by the firmware in the hardware configuration. If
CONFIG_ADSP_IDLE_CLOCK_GATING is enabled, this prevent is not active and
clock can be gated when core is in idle state. WIth this option disabled
clock gating will only be enabled in hardware during power gating.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -373,4 +385,35 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+
+	ARG_UNUSED(cnt);
+
+	__ASSERT(cnt >= 1, "Unbalanced clock gating lock get/put");
```
?? This should allows `0`

kv2019i replied:
Isn't this wrong. If we have just a single get and single put, "cnt" will be 0 zero and assert won't hold.

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -373,4 +385,35 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+{
+	if (atomic_get(&clock_gating_lock) == 0)
+		DSPCS.bootctl[arch_proc_id()].bctl |= DSPBR_BCTL_WAITIPCG;
+
```
shouldn't it be done only when the system is idle  ? (inside `pm_state_set()`)

kv2019i commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -373,4 +385,35 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+	if (atomic_get(&clock_gating_lock) == 0)
+		DSPCS.bootctl[arch_proc_id()].bctl |= DSPBR_BCTL_WAITIPCG;
+
+	atomic_inc(&clock_gating_lock);
```
For implementation, use of atomics is not bringing much added value here.  It does seem to be harmless and functionality will be ok still.

andyross replied:
I don't see any other synchronization, so these are needed.  It's a reference count that might be modified from racing contexts, so you can't do it with a non-atomic load/modify/store sequence.

kv2019i commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -373,4 +385,35 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+atomic_t clock_gating_lock;
+
+void adsp_clock_gating_lock_get(void)
+{
```
I wonder if we could have this get/put in a separate PR. 

I'm a bit worried how this going to be used. As it is here, this is very Intel specific function, so any app code using this, will be very specific to one type of hardware. OTOH, the substates were not much more generic in the end either. 

@tmleman In what case SOF would have to block clock gating? Is it just to implement the "prevent_local_clock_gating" and "prevent_power_gating" SET_D0IX IPC command fields? Or do we have some other needs? These IPCs are very Intel specific and for these the adsp_clock/power_gating_lock() interface is ok, but we need to put these into some Intel specific corner of SOF code as well.

@ceolin Do we have something in Zephyr like the QoS API in Linux, where a generic driver can set a generic constraint on responsiveness of the CPU, effectively blocking low-power states in a (still generic) manner.

tmleman replied:
> @tmleman In what case SOF would have to block clock gating? Is it just to implement the "prevent_local_clock_gating" and "prevent_power_gating" SET_D0IX IPC command fields? Or do we have some other needs? These IPCs are very Intel specific and for these the adsp_clock/power_gating_lock() interface is ok, but we need to put these into some Intel specific corner of SOF code as well.

Only for performance measurements. The motivation for me to propose such API was to give the ability to disable and enable this feature in the runtime from the SOF side without subsequent changes to the zephyr. But in the end it can be limit only to Kconfig.

kv2019i commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -373,4 +385,35 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+
+	__ASSERT(cnt >= 1, "Unbalanced clock gating lock get/put");
+
+	if (atomic_get(&clock_gating_lock) == 0)
```
So this might condition might be true twice if two threads race running lock_put(), but this is functionality ok as the bit gets clearer in the end.

kv2019i reviewed (commented):
The first part looks ok to enabling gating based on Kconfig. The adsp-specific interface to get/put a lock, that I'm less sure about and wonder whether this could be done in a separate PR (unless this lock capability is a must-have and we cannot enable anything without it). Comments inline.

andyross reviewed (commented):
nitpicks

tmleman commented:
I abandoned the desire to add an API that allows you to change clock gating settings in the runtime. Currently, the option to permanently enable this in the hardware is sufficient.

kv2019i reviewed (approved):
Thanks @tmleman , looks good! We can build on top of this to add runtime control later if/when needed (and have more time to discuss what is the proper interface to expose this).
